enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,751 @@
+    // - If the route specifies a HSTS annotation that matches one of the RequiredHSTSPolicies,
+    // a DomainPattern it specifies, but doesn't match a NamespaceSelector it specifies, the route is
+    // rejected with a message indicating the suggested policy value.
+    // - If the route specifies a HSTS annotation that doesn't match a
+    // RequiredHSTSPolicy, but matches a DomainPattern, the route is rejected with a message
+    // indicating the suggested policy value.
+    // - If the route specifies a HSTS annotation that doesn't match a
+    // RequiredHSTSPolicy, and also doesn't match a DomainPattern, the route is admitted.
+	// +optional
+	RequiredHSTSPolicies []RequiredHSTSPolicy `json:"requiredHSTSPolicies,omitempty"`
+}
+
+```
+The `route.openshift.io/RequiredRouteAnnotations` route validating admission plugin would follow the current OpenShift
+API server design pattern for admission plugins and additionally validate that routes matching
+the `RequiredHSTSPolicy` `NamespaceSelector` (if any) and `DomainPatterns`, are configured with the required HSTS
+policy.  If there are no `RequiredHSTSPolicies`, any route annotation will be valid.
+
+Some code is listed here, not all the trivial helper functions defined.
+````go
+...
+type requiredRouteAnnotations struct {
+	*admission.Handler
+	routeClient    routetypedclient.RoutesGetter
+	configClient   configtypedclient.IngressesGetter
+	nsLister       corev1listers.NamespaceLister
+	nsListerSynced func() bool
+}
+
+// Ensure that the required OpenShift admission interfaces are implemented.
+var _ = initializer.WantsExternalKubeInformerFactory(&requiredRouteAnnotations{})
+var _ = admissionrestconfig.WantsRESTClientConfig(&requiredRouteAnnotations{})
+var _ = admission.ValidationInterface(&requiredRouteAnnotations{})
+
+const hstsAnnotation = "haproxy.router.openshift.io/hsts_header"
+
+// Validate ensures that routes specify required annotations, and returns nil if valid.
+func (o *requiredRouteAnnotations) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+    ...
+    if o.getRequiredHSTS(ctx) == nil {
+        return nil
+    }
+    newRoute := a.GetObject().(*routeapi.Route)
+    ...
+    ingress, err = o.configClient.Ingresses().Get(ctx, "cluster", metav1.GetOptions{})
+    if err != nil {
+        return admission.NewForbidden(a, err)
+    }
+    namespace, err := o.nsLister.Get(newRoute.Namespace)
+    if err != nil {
+        return admission.NewForbidden(a, err)
+    }
+
+    if err := isRouteHSTSAllowed(ingress, newRoute, namespace); err != nil {
+        return admission.NewForbidden(a, err)
+    }
+    return nil
+}
+
+// isRouteHSTSAllowed returns nil if the route is allowed.  Otherwise, returns details and a suggestion in the error
+func isRouteHSTSAllowed(ingress *v1.Ingress, newRoute *routeapi.Route, namespace *corev1.Namespace) error {
+    // Invalid if a HSTS Policy is specified but this route is not TLS
```

Comment:
note that we cannot tighten validation generically unless this produced non-functional routes before.

Reply:
Ok, I will remove this.  I wanted to give early warning to the route author that HSTS is not allowed on non-TLS routes, but I can't do that and admit the route.  I can make this log a warning instead of invalidating the route.